### ISSUE TITLE
Visual Design for Keyboard Shorcuts Dialog

### ISSUE THREAD

**Issue by terracoda** [phetsims/balloons-and-static-electricity#263]
Content and rough visual layout for the keyboard shortcuts is available in a google doc [BASE:Keyboard Shortcuts](https://docs.google.com/document/d/1vk8_N3GuyC0o_Nl8HpBUu0tvVuWyWQ8FiksexPs_qBk/edit).

This issue is to finalize any content adjustments and visual layout.

I am assuming it would be nice to follow the look and feel of the dialog that was created for John Travoltage.

**Comment by emily-phet** [comment]
I've edited the document. I think it now uses the same wording as the dialog in John Travoltage. It also includes the arrow keys in all places the WASD keys are mentioned.
@arouinfar Can you make a mockup with this updated text?

**Comment by terracoda** [comment]
@emily-phet and @arouinfar I am reviewing the changes :-)

**Comment by arouinfar** [comment]
Can do @emily-phet, though I'm a bit tied up with the site visit. I should be able to have some mockups ready by Tuesday.

**Comment by emily-phet** [comment]
@arouinfar Thanks!

**Comment by terracoda** [comment]
@emily-phet, yes the addition of Shift key makes tiny steps when using both the arrow keys and the WASD keys.

**Comment by terracoda** [comment]
@emily-phet I made some changes to the content. Let me know what you think about these changes.

**Comment by emily-phet** [comment]
@terracoda I responded to your comments. Let me know what you think!

**Comment by terracoda** [comment]
I've updated the google doc and made a quick rough visual. I am sure @arouinfar can make it much nicer :-) 

I like the condensed version. It is based @emily-phet's feedback and edits, and carefully comparing to how the John Travoltage dialog is set up. The original content design was much more narrative to explain to interactions.

We can discuss order (top to bottom). I have put the Jump keys first, then Arrow keys because those are both new information for screen reader users. Screen reader users, get the WASD help text in the PDOM itself, but they do not get the JUMP keys or Arrow keys.

@arouinfar, let me know if you have any questions.

**Comment by arouinfar** [comment]
@terracoda thanks for your mockups! I've created a mockup in the JT style based on the condensed version. I've made the text a bit more succinct, but we can certainly wordsmith it a bit more. :)

![image](https://cloud.githubusercontent.com/assets/8419308/26125005/d0fbeec0-3a3d-11e7-9f18-30eac908e9c4.png)

**Comment by terracoda** [comment]
@arouinfar Wow! That looks nice, and I like the even shorter text. I do worry it's not explicit enough, which is why I struggled with getting it shorter, but we can have users determine if it is not clear. 

Do you think J plus S communicates that the user has to hold J and then press S to jump to sweater? 

I really like the way you combined the Move Grabbed Balloon section, super nice. 

The headings are all still very useful in terms of directing a blind user to the section they may be looking for as well. Nice!

**Comment by terracoda** [comment]
@arouinfar, the heading should be "Keyboard Shortcuts", and I think we should have a period at the end of every line. It helps with readability.

**Comment by arouinfar** [comment]
Thanks for the feedback @terracoda! I wasn't sure how explicit the text would need to be, but we can certainly adjust as necessary. 

> Do you think J plus S communicates that the user has to hold J and then press S to jump to sweater?

I do. I tried the commands in BASE, and for me I could simultaneous press J+S or hold S and then press J (or vice versa) so I think 'J plus S' should work. We also used similar text ('Shift plus Tab') to communicate how to move to the previously focused item.

I've added the missing periods and retitled the dialog (this is what happens when I mindlessly clone things in Illustrator :))
![image](https://cloud.githubusercontent.com/assets/8419308/26128121/932cf7c8-3a48-11e7-814a-c941fd7cad2c.png)

**Comment by terracoda** [comment]
@arouinfar, I think the only thing that might not be super clear is the Shift modifier key for taking smaller steps. We might need something more explicit there, but not sure. For example:

- Hold down __shift__ and press a drag key to move in smaller steps.

**Comment by terracoda** [comment]
Re previous comment https://github.com/phetsims/balloons-and-static-electricity/issues/263#issuecomment-302112166, the structural grouping may make it clear enough, just not 100%. I'll ask some screen reader users.

**Comment by arouinfar** [comment]
> Hold down shift and press a drag key to move in smaller steps.

Excellent suggestions @terracoda! 

![image](https://cloud.githubusercontent.com/assets/8419308/26181797/25111c72-3b30-11e7-9301-7acf4db6160b.png)

**Comment by terracoda** [comment]
@arouinfar, I think this is looking really good. I'll mock it up in HTML next.

**Comment by terracoda** [comment]
@arouinfar, how about we add _up, left, down, or right_ to the WASD key description, so it looks like this:
> __W, A, S, or D keys__ drag and rub grabbed balloon up, left, down, or right.

Hopefully this doesn't go to 3 lines.

**Comment by terracoda** [comment]
@arouinfar, instead of creating an static HTML page, I updated the sim code #266.

**Comment by arouinfar** [comment]
@terracoda I'm not sure how to see the changes you've made in #266 since it looks like they're not in master yet, but here's another screenshot
![image](https://cloud.githubusercontent.com/assets/8419308/26212237/25904bb4-3bb2-11e7-8c3d-4a881fb473ba.png)

**Comment by terracoda** [comment]
@jessegreenberg, @arouinfar, @emily-phet I also made an [html version](http://terracoda.net/research/phet-examples/base-keyboard-shortcuts.html) in case we want to test the content out of context for interpretation.

**Comment by terracoda** [comment]
@arouinfar, do you think we should capitalize all teh instances of a key name? "Enter", "Space", etc?

**Comment by terracoda** [comment]
@arouinfar, I updated the google doc content and made an HTML version that I posted on my website above.

We have to wait for @jessegreenberg to review the changes and merge them.

**Comment by terracoda** [comment]
@arouinfar, I think it looks really good, the only keys needing a capital letters is the "Enter" in the first sentence, and "Shift" in the smaller steps sentence.

**Comment by jessegreenberg** [comment]
I just merged the changes into master. And thanks for the HTML version, that is helpful!

**Comment by terracoda** [comment]
@jessegreenberg, I think the only difference in structure between the HTML version and the PhET code version is the list structure under the **Move Grabbed Balloon** heading. In the PhET version I left the three items as paragraphs as I didn't want to mess up the code too much, but I think a list makes sense.

**Comment by jessegreenberg** [comment]
Agreed, a list sounds great to me. In https://github.com/phetsims/balloons-and-static-electricity/issues/263#issuecomment-302462826 the mockup looks great, but the spacebar icon looks a little strange to me.  Could we add "Space" or "Spacebar" to that icon?

**Comment by terracoda** [comment]
@jessegreenberg, I am good with adding "Space."

**Comment by terracoda** [comment]
@arouinfar, we got some great feedback from the A11y Meeting today.
First a small correction to the text which I missed.
"J plus N jumps close to wall." should be "J plus N jumps near wall." The N is for "Near".

Feedback from A11y Group:
- Make space key make wider and/or less tell.
- Change to use “move” instead of drag and rub.
- Bold letters and words on keys.
- Bold destinations on jump hotkeys and unbold key wording.
- Adjust/play with alignment of arrow/WASD text blocks:
  - could we see versions with the blocks middle aligned on top key and on bottom row?

**Comment by terracoda** [comment]
@arouinfar, I forgot one item. 
- use "or" instead of slash in first sub-heading "Grabs / Releases Balloon."

I've updated the [design document](https://docs.google.com/document/d/1vk8_N3GuyC0o_Nl8HpBUu0tvVuWyWQ8FiksexPs_qBk/edit) and the [HTML sample](https://github.com/terracoda/phet-templates/blob/master/base-keyboard-shortcuts.html).

**Comment by arouinfar** [comment]
Thanks for the feedback @terracoda! Here're some updated mockups with various alignments in the arrow/WASD groups, sorted in order of my personal preference.

![image](https://cloud.githubusercontent.com/assets/8419308/26470875/26a18aac-4154-11e7-90d8-12aa5b73af43.png)
![image](https://cloud.githubusercontent.com/assets/8419308/26470882/2f8679d4-4154-11e7-8dab-6511b8e0db00.png)
![image](https://cloud.githubusercontent.com/assets/8419308/26470894/3c90c846-4154-11e7-9982-e18e23085015.png)
![image](https://cloud.githubusercontent.com/assets/8419308/26470901/429def16-4154-11e7-9c07-a21b75d777c1.png)
![image](https://cloud.githubusercontent.com/assets/8419308/26470906/48e6183a-4154-11e7-82c0-1fd54ea41702.png)

**Comment by jessegreenberg** [comment]
These look great @arouinfar, I personally really like the last one "center-aligned to bottom row".

**Comment by terracoda** [comment]
@emily-phet, I updated the [google design doc](https://docs.google.com/document/d/1vk8_N3GuyC0o_Nl8HpBUu0tvVuWyWQ8FiksexPs_qBk/edit#) to represent 3 very similar versions of the content that we discussed today:
I updated the bolding to what I think we agreed upon and then adjusted the content to have no mention of keys, and one to have heading structure added for the move section.
1. Super condensed version
2. Super condensed version "no keys"
3. Move Grabbed Balloon section with subheads

@emily-phet, which one do you think works best? or do we need to mix and match a bit?

**Comment by terracoda** [comment]
@emily-phet, I think I prefer sticking with the word "key" because that implies you have to press a key. Without it, I feel we should be saying "Press" somewhere, especially for the WASD key section.

I we added a general heading for "Move Grabbed Balloon" then we might want to go back to using the verb "drag" to describe the "normal" move.

**Comment by terracoda** [comment]
I updated the [html here in my repo](https://github.com/terracoda/phet-templates/blob/master/base-keyboard-shortcuts.html), but haven't published the new version.

**Comment by emily-phet** [comment]
@arouinfar - Thanks! This was very helpful.
In a design meeting today, we decided to move forward with the "center-aligned to bottom row" version. Can you take this version and update the design to include:
- double the padding between the W,A,S,D keys and associated text, and the shift (and associated text) below. 
- text formatting and wording changes to the text as indicated at the top of this [design doc](https://docs.google.com/document/d/1vk8_N3GuyC0o_Nl8HpBUu0tvVuWyWQ8FiksexPs_qBk/edit#).

**Comment by emily-phet** [comment]
@terracoda 
- I updated your design doc (with a new section) with wording changes and bolding/underlining as we discussed today. This section has a heading indicating that this is the one for Amy to work on implementing.
- Regarding the use of the word "key" - I came to the same conclusion you did...when I took it out I didn't like it...so I put it back in. The version for Amy to implement includes the words key/keys as before.
- Regarding the html, do you know what that is used for? Does @jessegreenberg use this directly?

**Comment by terracoda** [comment]
@emily-phet, yes, the HTML version is for @jessegreenberg. It has all the structure. I will make sure it has the same content as your updated version.

**Comment by terracoda** [comment]
@emily-phet, @jessegreenberg , @arouinfar I updated the HTML in [phet-templates/base-keyboard-shortcuts.html](https://github.com/terracoda/phet-templates/blob/master/base-keyboard-shortcuts.html). Here's a screenshot of the rendered page. 

@emily-phet, I'm just wondering if we need a comma after the  jump keys?
"J plus S, Jumps close to Sweater."
![base-keyboard-shorcuts](https://cloud.githubusercontent.com/assets/1902116/26525473/90459f68-4332-11e7-8b5e-3aa6d63a15aa.png)

**Comment by arouinfar** [comment]
@emily-phet @terracoda @jessegreenberg I've created an updated mockup based on the feedback in https://github.com/phetsims/balloons-and-static-electricity/issues/263#issuecomment-304374994. I doubled the amount vertical padding between the items in the "Move Grabbed Button" section. Please let me know if that wasn't what was requested. 

![image](https://cloud.githubusercontent.com/assets/8419308/26650129/af10f69c-4605-11e7-965b-1f6bea4116b7.png)

**Comment by terracoda** [comment]
@arouinfar, I think @emily-phet meant to put space below the WASD keys only, i.e., above the Shift key. The small white-space above the WASD keys can remain small, consistent with the rest of the spacing because the blocks of text are naturally separated due to the arrangement of the keys.

Otherwise, looks awesome.

**Comment by arouinfar** [comment]
Thanks for the clarification @terracoda! Here are two options with increased spacing between the WASD group and the Shift.

![image](https://cloud.githubusercontent.com/assets/8419308/26689013/82447484-46b1-11e7-93f6-35ea2a157372.png)

![image](https://cloud.githubusercontent.com/assets/8419308/26688976/657f97c0-46b1-11e7-8c12-31e088fd46d8.png)

**Comment by terracoda** [comment]
@arouinfar, I prefer the second with more space.

**Comment by arouinfar** [comment]
> @arouinfar, I prefer the second with more space.

Me too, @terracoda!

**Comment by emily-phet** [comment]
@arouinfar Looks great! I like the one with double padding. I think we're good to go here.

**Comment by terracoda** [comment]
Great! Thanks everyone! Updating #258 checklist.

**Comment by jessegreenberg** [comment]
Removing assignment until we come to a conclusion about what design for keyboard help dialog should be.

**Comment by terracoda** [comment]
Assigned myself to keep it on my list. We will make decision after I review things for next week's Design Crit.

**Comment by terracoda** [comment]
@arouinfar, there were some great suggestions from the Design Crit. One was this visually connected version from Caren Watkins. It connects the grab and release functions and multiple move functions. This visual connection may communicate the custom two-step interaction more clearly. I am posting it here for reference as it maybe nice to consider.
﻿
<img width="650" alt="base-kbdialog" src="https://user-images.githubusercontent.com/1902116/31947033-febc9fc8-b8ad-11e7-8b13-2b5f3706086c.png">

And I will start a new issue with the draft content of a compact version for the BASE keyboard shortcuts dialog. Not sure if the above layout idea still applies in the compact version, but it might.

**Comment by jessegreenberg** [comment]
@terracoda can this issue be closed? I think we are continuing in #320?

**Comment by terracoda** [comment]
@jessegreenberg, I mentioned this issue on #320, so I think it is ok to close this one. Closing now.
